pkgs/development/haskell-modules/make-package-set.nix

```diff
@@ -315,6 +315,245 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
           else "${ghcEnv}/lib/${ghcCommand}-${ghc.version}";
       });
 
+    # This uses stack2nix to create a .nix file corresponding to a Haskell
+    # package set for a Stack project.  This .nix file can either be copied
+    # into your project and used as-is, or it can be used directly with the
+    # callStack2nixPkgSet function.
+    #
+    # This function is meant for easily creating a package set for building a
+    # Stack-based project.
+    #
+    # Here's an example of how to use this:
+    #
+    #   haskellPackages.callStack2nix {
+    #     name = "purescript";
+    #     src = fetchFromGitHub {
+    #       owner = "purescript";
+    #       repo = "purescript";
+    #       rev = "v0.12.4";
+    #       sha256 = "0lgfmqrn2j1fipgghxrccd9n5f8sw1ncc7fb25201z6x3klzy7bb";
+    #     };
+    #     sha256 = "1in80b914h46hsscdvkmfz0042181lxy2cj522gzyc36l4qq560w";
+    #     hackageSnapshotTimestamp = "2019-05-06T00:00:00Z";
+    #     compiler = haskell.compiler.ghc864;
+    #   };
+    #
+    # This creates a .nix file containing a Haskell package set with Haskell
+    # package versions determined by the LTS used to build PureScript.  This
+    # example is using LTS-13.12 (which is determined by the stack.yaml file in
+    # the purescript repo).
+    #
+    # See the callStack2nixPkgSet for how to import and build this resulting
+    # package set.
+    #
+    # In general, this function won't be used within nixpkgs, but instead
+    # by end-users who want an easy way to build a stack-based project with nix.
+    #
+    # This produces a fixed-output derivation.  See the comment on the sha256
+    # argument for a little more info about this.  This means that
+    # callStack2nix is suited for building Haskell tools, but not suited for
+    # interactive development on Haskell projects.
+    #
+    # For instance, with the PureScript example above, callStack2nix is
+    # convenient to use if you're working on a frontend project written in
+    # PureScript, and you need the PureScript compiler available in your
+    # environment.  However, callStack2nix should not be used if you want to
+    # hack on the PureScript compiler itself.
+    #
+    # TODO: Currently this doesn't work for stack.yaml files that have extra-deps
```

Comment:
@cdepillabout what is the unfortunate interaction? I'd be interested to using your work but need to address this restriction :)

Reply:
I didn't look at this super deeply, but I think what happens is the following:

1. `stack2nix` calls into the `stack` codebase to do a bunch of stuff.  Part of this process involves `stack` downloading all of the `extra-deps` defined in the `stack.yaml` that are specified as URLs, git repos, etc.  These packages are downloaded to `.stack-work/downloaded/`.
2. `stack2nix` then calls into the `cabal2nix` codebase to actually create nix derivations for each Haskell package defined in the stackage LTS, as well as all the packages defined in `extra-deps`.
3. For `extra-deps` that can be fetched from Hackage, I think `cabal2nix` just downloads the .cabal file for the package and runs on that.  For packages that can't be fetched from Hackage (for instance, Haskell packages that are only available on Github), `cabal2nix` downloads the repo using tools like `nix-prefetch-git`, and then runs on the .cabal file contained in the repo.  However, using prefetchers like `nix-prefetch-git` doesn't work when running under `nix-build`.

----------------------------------------------------------------------------------------------------

Ideally, I'd like to get this PR merged in, since it does still work in many instances (specifically, for any stack project that doesn't contain Git repos as `extra-deps`).  After that, I'd like to send a PR to the `stack2nix` repo fixing it up so that it doesn't cause `cabal2nix` to try to call tools like `nix-prefetch-git`.

Reply:
I think a better way is to add support to all those prefetchers not to use `nix-store --add` as populating nix store isn't really needed.

I'm not entirely sure to merge this before that's done, as it will give a lot of failed builds and bug reports. I do see it brings value to you though, I'm a bit undecisive to be honest.

Reply:
> I think a better way is to add support to all those prefetchers not to use nix-store --add as populating nix store isn't really needed.

I don't know if this change would be accepted to those prefetchers, since putting stuff into `/nix/store` is basically why they exist.

However, I think you're on the right track.  It might be a good idea to open up an issue against `cabal2nix` that stops it from using the prefetchers, and instead just downloads things directly with tools like `git`.  There is no reason for it to need to use the prefetchers, aside from the fact that it makes the prefetchers make it easy to calculate the sha256 hash.  Although I don't know how easy it would be to get this change into `cabal2nix`.  It may be simpler to just work around it in `stack2nix`.

> I'm not entirely sure to merge this before that's done, as it will give a lot of failed builds and bug reports.

Yeah, I think I agree with you here.  I think we need to determine whether we value having this (somewhat?) useful function in nixpkgs, even though we know it won't work for some people, or whether we value reducing bug reports.

When working on this PR, I searched for Haskell repos online I could test it on, and I actually couldn't find even one that used non-Hackage `extra-deps` (granted, I only searched for about 5 minutes).  To me, this pushes it over the line of usefulness, so I'm slightly in favor of merging it in.

But I definitely agree that it would be ideal to have it working fully in all cases before merging it in.

Reply:
> I don't know if this change would be accepted to those prefetchers, since putting stuff into /nix/store is basically why they exist.

Maybe we could have "hash generators" that prefetchers would use, but I'm not entirely sure that it's worth the effort than just adding something like `--no-store-add` toggle to existing prefetchers.

> I think we need to determine whether we value having this (somewhat?) useful function in nixpkgs

Maybe with revised documentation that clearly states the limitations (I had to ask a question why this TODO means), it would add value while documenting where it won't work. Bonus would be actually parsing `extra-deps` and check if it uses anything else than cabal package and fail with a good error. But then energy is better spent just getting fetchers to be more generic :)

Reply:
> Maybe we could have "hash generators" that prefetchers would use, but I'm not entirely sure that it's worth the effort than just adding something like --no-store-add toggle to existing prefetchers.

I created an issue about either having a hash generator, or adding a `--no-store-add` flag to the `nix-prefetch-url` command:

https://github.com/NixOS/nix/issues/2919

I haven't updated the documentation yet on this PR, but I will let you know when I do.

Reply:
@domenkozar I just added a commit that changes the documentation to explain in more detail when `callStack2nix` cannot be used currently:

2c40da7